File: src/01.cpp
#include <iostream>
#include <vector>

#include "common/common.h"

const int HUGE = 99999;

int solve_1(const std::vector<int> &numbers) {
    int previous = HUGE;
    int result = 0l;
    for (const int n : numbers) {
        if (n > previous) {
            ++result;
        }
        previous = n;
    }
    return result;
}

int solve_2(const std::vector<int> &numbers) {
    int p_1 = HUGE, p_2 = HUGE, p_3 = HUGE;
    int result = 0l;
    for (const auto n : numbers) {
        if (n + p_1 + p_2 > p_1 + p_2 + p_3) {
            ++result;
        }
        p_3 = p_2;
        p_2 = p_1;
        p_1 = n;
    }
    return result;
}

int main(int argc, char** argv) {
    parse_fn<int> parse = [](const std::string &s) {return std::stoi(s);};
    const auto numbers = getParsedLines(argv[1], parse);
    std::cout << solve_1(numbers) << std::endl;
    std::cout << solve_2(numbers) << std::endl;
}

File: src/01.rs
#[path = "common/common.rs"]
mod common;

use common::get_parsed_lines;

const HUGE: i32 = 9999999;

fn solve_1(numbers: &Vec<i32>) -> i32 {
    let mut previous = HUGE;
    let mut result = 0;
    for &n in numbers {
        if n > previous {
            result += 1;
        }
        previous = n;
    }
    result
}

fn solve_2(numbers: &Vec<i32>) -> i32 {
    let (mut p_1, mut p_2, mut p_3) = (HUGE, HUGE, HUGE);
    let mut result = 0;
    for &n in numbers {
        if n + p_1 + p_2 > p_1 + p_2 + p_3 {
            result += 1;
        }
        (p_1, p_2, p_3) = (n, p_1, p_2);
    }
    result
}

fn main() {
    let numbers = get_parsed_lines(|s| s.parse().unwrap());
    println!("{}", solve_1(&numbers));
    println!("{}", solve_2(&numbers));
}
